### PR TITLE
Prevent upload (overwrite) of lfs locked file

### PR DESCRIPTION
**Pull request by davidsvantesson**
Fix one point of #8687.

### CODE REVIEW

File: modules/repofiles/upload.go [hunk 1]
```diff
@@ -55,6 +55,23 @@ func UploadRepoFiles(repo *models.Repository, doer *models.User, opts *UploadRep
...
 	}
 
+	names := make([]string, len(uploads))
+	infos := make([]uploadInfo, len(uploads))
+	for i, upload := range uploads {
+		names[i] = upload.Name
+		infos[i] = uploadInfo{upload: upload}
+
+		// Check file is not lfs locked
+		filepath := path.Join(opts.TreePath, upload.Name)
```
Comment (lunny) [inline modules/repofiles/upload.go:62]: Move these before line 61?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
And we also need to change HTTP request?

**Comment by davidsvantesson** [comment]
> And we also need to change HTTP request?

Can you elaborate on this comment?

File: modules/repofiles/upload.go (hunk 1 above)
Comment (zeripath) [reply modules/repofiles/upload.go:62]: I presume that this was referring to some other lines which have been moved? (presumably lines 58 & 59)

As `filepath` has to be set within the loop as it uses `upload.Name`

File: modules/repofiles/upload.go [hunk 2]
```diff
@@ -55,6 +55,23 @@ func UploadRepoFiles(repo *models.Repository, doer *models.User, opts *UploadRep
 		return fmt.Errorf("GetUploadsByUUIDs [uuids: %v]: %v", opts.Files, err)
 	}
 
+	names := make([]string, len(uploads))
+	infos := make([]uploadInfo, len(uploads))
+	for i, upload := range uploads {
+		// Check file is not lfs locked
+		filepath := path.Join(opts.TreePath, upload.Name)
+		lfsLock, err := repo.GetTreePathLock(filepath)
```
Comment (zeripath) [inline modules/repofiles/upload.go:63]: It may be that we would be better off batching these in some way - however, that could be an early optimisation.

Comment (davidsvantesson) [reply modules/repofiles/upload.go:63]: I think the normal case would be to only upload a single or maybe a few files.

Comment (zeripath) [inline modules/repofiles/upload.go:63]: Sorry I missed this earlier - you should only check LFS Locks if `setting.LFS.Server`. You need to wrap this code in a check.

**Review by zeripath — CHANGES_REQUESTED** [review]

File: modules/repofiles/upload.go (hunk 2 above)
Comment (davidsvantesson) [reply modules/repofiles/upload.go:63]: It's actually in GetTreePathLock, but I am not sure if it is a good design.

Comment (zeripath) [reply modules/repofiles/upload.go:63]: Ah ok just add a comment above this line to say that if LFS is off then this will return nil.

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: modules/repofiles/upload.go (hunk 2 above)
Comment (davidsvantesson) [reply modules/repofiles/upload.go:63]: There should probably be a code refactoring regarding GetTreePathLock, but I went for the easy way now and only added a code comment.

**Pull request merged**